lib/charms/loki_k8s/v0/charm_logging.py

```diff
@@ -0,0 +1,463 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""This charm library contains utilities to automatically forward your charm logs to a loki-push-api endpoint.
+
+(yes! charm code, not workload code!)
+
+This means that, if your charm is related to, for example, COS' Loki charm (or a Grafana Agent),
+you will be able to inspect in real time from the Grafana dashboard the logs emitted by your charm.
+
+To start using this library, you need to do two things:
+1) decorate your charm class with
+
+`@log_charm(loki_push_api_endpoint="my_logging_endpoint")`
+
+2) add to your charm a "my_logging_endpoint" (you can name this attribute whatever you like) **property**
+that returns an http/https endpoint url. If you are using the `LogProxyConsumer` as
+`self.logging = LogProxyConsumer(self, ...)`, the implementation could be:
+
+```
+    @property
+    def my_logging_endpoint(self) -> List[str]:
+        '''Loki push API endpoints for charm logging'''
+        return self.logging.loki_endpoints:
+```
+
+The ``log_charm`` decorator will take these endpoints and set up the root logger (as in python's
+logging module root logger) to forward all logs to these loki endpoints.
+
+3) If you were passing a certificate using `server_cert`, you need to change it to provide an *absolute* path to
+the certificate file.
+"""
+import copy
+import functools
+import logging
+import os
+import string
+import time
+from contextlib import contextmanager
+from logging.config import ConvertingDict
+from pathlib import Path
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Optional,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+    cast,
+)
+
+import requests
+from ops.charm import CharmBase
+from ops.framework import Framework
+
+# prevent infinite recursion because on failure urllib3 will push more logs
+# https://github.com/GreyZmeem/python-logging-loki/issues/18
+logging.getLogger("urllib3").setLevel(logging.INFO)
+
+# The unique Charmhub library identifier, never change it
+LIBID = "52ee6051f4e54aedaa60aa04134d1a6d"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+PYDEPS = []
+
+logger = logging.getLogger("charm_logging")
```

Comment:
This is a neat capability, well done!
I wonder if we could simplify this:
In charm.py we always have `logger = logging.getLogger(__name__)`. Would it make sense / be possible to somehow update this logger, or create a mirror one, that sends to loki? I.e. without a decorator.

Reply:
The thing is, and it's the same with `charm_tracing`, that the charm instance (not the type, nor the context) know what endpoint to send the data to.
So we first need to initialize the charm, then we need to ask the charm 'what endpoint do you want me to send telemetry to', and then we can send stuff.

We _could_ do all of this without the charm instance knowing about it, but we'd need to (hard)code plenty of raw hook tool calls to figure out what the endpoint is. 
For example say we're instrumenting traefik. We want to set up a logger to send logs to a loki instance. The endpoint is in the 'logging' endpoint relation databag. 
We'd need to, in charm.py:

```
charm_logging.start(
loki_endpoint = Popen("relation-get <ID of the relation on 'logging' if present>").from_json()["endpoint"]["url"]
)
```

And then we're not even considering TLS. It gets pretty complex to do this without the API the charm has access to. 
[Which is one more argument for separating ops API and charm API as far as I'm concerned.
](https://github.com/canonical/operator/issues/1075)

Reply:
TLDR: that is possible but IMHO not desirable and too complex, It would require a redesign of charm tracing and charm logging and further speccing work. Not something for this PR.